cloud/store/recover.go

```diff
@@ -0,0 +1,192 @@
+		var decodedData map[string]interface{}
+		err := msgpack.Unmarshal(log.Data, &decodedData)
+		if err != nil {
+			st.log.Error("unable to unmarshal msgpack from raft.db file err=" + err.Error())
+			continue
+		}
+
+		// Extract the "Servers" array from the map
+		servers, ok := decodedData["Servers"].([]interface{})
+		if !ok {
+			st.log.Error("failed to extract 'Servers' array")
+			continue
+		}
+
+		// Iterate over the array and decode each Peer
+		for _, server := range servers {
+			serverMap, ok := server.(map[string]interface{})
+			if !ok {
+				st.log.Error("failed to convert server to map")
+				continue
+			}
+
+			var peer Peer
+			if address, ok := serverMap["Address"].(string); ok {
+				peer.Address = address
+			}
+			if id, ok := serverMap["ID"].(string); ok {
+				peer.ID = id
+			}
+			if suffrage, ok := serverMap["Suffrage"].(int64); ok {
+				peer.Suffrage = suffrage
+				peer.NonVoter = suffrage != int64(raft.Voter)
+			}
+
+			existedP := peersMap[peer.Address]
+			if reflect.DeepEqual(existedP, peer) {
+				continue
+			}
+			peersMap[peer.Address] = peer
+			peers = append(peers, peer)
+		}
+	}
+
+	if len(peers) == 0 {
+		return nil
+	}
+
+	configData, err := json.MarshalIndent(peers, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	file, err := os.OpenFile(peerFilePath, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0o644)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	if _, err := file.Write(configData); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// configViaPeers checks if the peers.json file exists with the provided path
+// if the files exists it creates Raft.Configuration from it
+func (st *Store) configViaPeers(peerFilePath string) (raft.Configuration, error) {
+	var configuration raft.Configuration
+	if _, err := os.Stat(peerFilePath); err != nil {
+		st.log.Error("either file doesn't exist, or there an error", err)
+		// if files doesn't exist we bail out early
+		return configuration, nil
+	}
+
+	st.log.Info(fmt.Sprintf("found %s file, collect peers from previous config...", peerFilePath))
+
+	configuration, err := raft.ReadConfigJSON(peerFilePath)
+	if err != nil {
+		return configuration, fmt.Errorf("raft recovery failed to parse peers.json: %w", err)
+	}
+	if err := os.RemoveAll(peerFilePath); err != nil {
+		return configuration, fmt.Errorf("recovery failed to delete %s, please delete manually ,err=%w", peerFilePath, err)
+	}
+
+	st.log.Info(fmt.Sprintf("deleted %s file and will proceed with the recovery", peerFilePath))
+
+	st.log.Info("previous raft", "peers", fmt.Sprintf("%v", configuration.Servers))
+	return configuration, nil
+}
+
+// TODO-WEAVIATE-680: at the moment the validation of the provided address happens on the ip level but not ports,
+// improvement would be validate also that the provided ip and ports exists in the memberlist.
+func (st *Store) recoverable(servers []raft.Server, sleepTime time.Duration) ([]raft.Server, bool) {
+	if len(servers) == 0 {
```

Comment:
The nodes will not be instantly available, so I think it makes sense to give some time for the peers to join and retry

Reply:
it's the case [here](https://github.com/weaviate/weaviate/pull/4087/files#diff-71a484f3dccfe2199cc13272f254d716ea12cdaf27ee8e29ef66ae5d5717f2abR171), that is the length of the one provided from the peers file if it had empty list

Reply:
Sure here it is just waiting once at the beginning. Ideally, we should retry a couple of times before giving up

Reply:
the `time.Sleep ` is configurable so i think we can extend it via config, or I am missing something ?

Reply:
recoverable() is just called once and blocks for sleepTime.
A better approach is to retry a couple of times, not just once, and return if the context expires.

Reply:
what could change between `retry multiple times` and `sleep for sometime` with the current approach 🤔  ?

Reply:
Sure. Here's an example that may clarify the matter: let's say sleepTime = 1 * time.Minute. In this case, the app will block for 1 minute, even though all the servers the node is looking for already exist or are about to join in a few seconds.